packages/astro/src/i18n/middleware.ts

```diff
@@ -54,30 +55,52 @@ export function createI18nMiddleware(
 
 		if (response instanceof Response) {
 			const pathnameContainsDefaultLocale = url.pathname.includes(`/${defaultLocale}`);
-			if (i18n.routing === 'prefix-other-locales' && pathnameContainsDefaultLocale) {
-				const newLocation = url.pathname.replace(`/${defaultLocale}`, '');
-				response.headers.set('Location', newLocation);
-				return new Response(null, {
-					status: 404,
-					headers: response.headers,
-				});
-			} else if (i18n.routing === 'prefix-always') {
-				if (url.pathname === base + '/' || url.pathname === base) {
-					if (trailingSlash === 'always') {
-						return context.redirect(`${appendForwardSlash(joinPaths(base, i18n.defaultLocale))}`);
-					} else {
-						return context.redirect(`${joinPaths(base, i18n.defaultLocale)}`);
+			switch (i18n.routing) {
+				case 'pathname-prefix-other-locales': {
+					if (pathnameContainsDefaultLocale) {
+						const newLocation = url.pathname.replace(`/${defaultLocale}`, '');
+						response.headers.set('Location', newLocation);
+						return new Response(null, {
+							status: 404,
+							headers: response.headers,
+						});
+					}
+					break;
+				}
+
+				case 'pathname-prefix-always-no-redirect': {
+					// We return a 404 if:
+					// - the current path isn't a root. e.g. / or /<base>
+					// - the URL doesn't contain a locale
+					const isRoot = url.pathname.startsWith(base) || url.pathname === '/';
```

Comment:
In line 86 below, it's using `url.pathname === base + '/' || url.pathname === base` to check the root. Is that more accurate? Since the `startsWith` here could match against deep paths.

Reply:
Yeah, it makes sense